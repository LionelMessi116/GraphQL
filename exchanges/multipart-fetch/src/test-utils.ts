import {
  GraphQLRequest,
  OperationContext,
  Operation,
  makeOperation,
} from '@urql/core';
import gql from 'graphql-tag';

const context: OperationContext = {
  fetchOptions: {
    method: 'POST',
  },
  requestPolicy: 'cache-first',
  url: 'http://localhost:3000/graphql',
};

const queryGql: GraphQLRequest = {
  key: 2,
  query: gql`
    query getUser($name: String) {
      user(name: $name) {
        id
        firstName
        lastName
      }
    }
  `,
  variables: {
    name: 'Clara',
  },
};

const obj = { hello: 'world' };
const file = new File(
  [new Blob([JSON.stringify(obj, null, 2)], { type: 'application/json' })],
  'index.ts'
);
const files = [file, file];

const upload: GraphQLRequest = {
  key: 3,
  query: gql`
    mutation uploadProfilePicture($picture: File) {
      uploadProfilePicture(picture: $picture) {
        location
      }
    }
  `,
  variables: {
    picture: file,
  },
};

const uploads: GraphQLRequest = {
  key: 3,
  query: gql`
    mutation uploadProfilePictures($pictures: [File]) {
      uploadProfilePicture(pictures: $pictures) {
        location
      }
    }
  `,
  variables: {
    picture: files,
  },
};

export const uploadOperation: Operation = makeOperation(
  'mutation',
  upload,
  context
);

export const multipleUploadOperation: Operation = makeOperation(
  'mutation',
  uploads,
  context
);

export const queryOperation: Operation = makeOperation(
  'query',
  queryGql,
  context
);
