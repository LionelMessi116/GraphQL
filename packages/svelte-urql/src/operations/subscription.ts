import { OperationContext, CombinedError, createRequest } from '@urql/core';
import { pipe, fromValue, concat, scan, map, subscribe } from 'wonka';
import { Readable } from 'svelte/store';
import { DocumentNode } from 'graphql';

import { getClient } from '../context';
import { initialState } from './constants';

export interface SubscriptionArguments<V> {
  query: string | DocumentNode;
  variables?: V;
  context?: Partial<OperationContext>;
}

export type SubscriptionHandler<T, R> = (prev: R | undefined, data: T) => R;

export interface SubscriptionResult<T> {
  fetching: boolean;
  stale: boolean;
  data?: T;
  error?: CombinedError;
  extensions?: Record<string, any>;
}

export const subscription = <T = any, R = T, V = object>(
  args: SubscriptionArguments<V>,
  handler?: SubscriptionHandler<T, R>
): Readable<SubscriptionResult<T>> => {
  const client = getClient();
  const request = createRequest(args.query, args.variables as any);

  const queryResult$ = pipe(
    concat([
      fromValue({ fetching: true, stale: false }),
      pipe(
        client.executeSubscription(request, args.context),
        map(({ stale, data, error, extensions }) => ({
          fetching: false,
          stale: !!stale,
          data,
          error,
          extensions,
        }))
      ),
      fromValue({ fetching: false, stale: false }),
    ]),
    scan((result, partial: any) => {
      const data =
        partial.data !== undefined
          ? typeof handler === 'function'
            ? handler(result.data, partial.data)
            : partial.data
          : result.data;
      return { ...result, ...partial, data };
    }, initialState)
  );

  return {
    subscribe(onValue) {
      const { unsubscribe } = pipe(queryResult$, subscribe(onValue));
      return unsubscribe as () => void;
    },
  };
};
