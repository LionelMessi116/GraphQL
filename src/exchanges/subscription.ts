import { parse } from 'graphql/language/parser';
import { merge, Observable, Observer, of } from 'rxjs';
import { mergeMap, partition } from 'rxjs/operators';
import { Exchange, ExchangeResult, Operation } from '../types';

function extractOperationName(operation: Operation): Operation {
  const doc = parse(operation.query);

  // TODO: How do we enforce this? I feel like this has to be stable given the spec?
  // this is grabbing the first selection from the subscription query. Which is required to only be 1.
  // and backend servers require operationName to be the subscription name
  const operationName =
    // @ts-ignore
    doc.definitions[0].selectionSet.selections[0].name.value;

  return {
    ...operation,
    operationName,
  };
}
const liveSubscriptions = new Map<string, () => void>();

export const subscriptionExchange: Exchange = ({ forward }) => {
  const handleSubscription = (operation: Operation) => {
    if (operation.context.unsubscribe) {
      return new Observable<ExchangeResult>(observer => {
        const unsubscribe = liveSubscriptions.get(operation.query);
        // clear away queries
        liveSubscriptions.delete(operation.query);

        if (unsubscribe) {
          unsubscribe();
        }

        observer.complete();
      });
    }

    return new Observable<ExchangeResult>(observer => {
      const { unsubscribe } = operation.context.forwardSubscription(
        extractOperationName(operation),
        observer
      );

      liveSubscriptions.set(operation.query, unsubscribe);
    });
  };

  return ops$ => {
    return ops$.pipe(
      source$ =>
        of(
          partition<Operation>(op => op.operationName === 'subscription')(
            source$
          )
        ),
      // @ts-ignore
      mergeMap(([subscriptions$, operations$]) =>
        merge(
          subscriptions$.pipe(mergeMap(handleSubscription)),
          operations$.pipe(forward)
        )
      )
    );
  };
};
