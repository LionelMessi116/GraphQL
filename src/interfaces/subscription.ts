// These are *not* Observable primitives, but are instead intended to
// be used with the subscriptionExchange

export interface ISubscriptionObserver {
  next: (IExecutionResult) => void;
  error: (Error) => void;
}

export interface ISubscription {
  unsubscribe: () => void;
}
