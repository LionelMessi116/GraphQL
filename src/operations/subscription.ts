import { OperationContext, createRequest } from 'urql/core';
import { DocumentNode } from 'graphql';
import { getClient } from '../context';
import { observe } from './observe';

export interface SubscriptionArguments<Variables> {
  query: DocumentNode | string;
  variables?: Variables;
  pause?: boolean;
  context?: Partial<OperationContext>;
}

export const subscribe = <Variables = object>(
  args: SubscriptionArguments<Variables>
) => {
  const client = getClient();
  const request = createRequest(args.query, args.variables as any);

  return observe(client.executeSubscription(request, args.context));
};
