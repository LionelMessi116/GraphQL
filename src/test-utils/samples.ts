import {
  ExchangeResult,
  ExecutionResult,
  Mutation,
  Operation,
  Query,
  Subscription,
} from '../types';

const context = {
  fetchOptions: {
    method: 'POST',
  },
  url: 'http://localhost:3000/graphql',
};

export const queryGql: Query = {
  query: `query getUser($name: String){
    user(name: $name) {
      id
      firstName
      lastName
    }
  }`,
  variables: {
    name: 'Clara',
  },
};

export const mutationGql: Mutation = {
  query: `mutation AddUser($name: String){
    addUser(name: $name) {
      name
    }
  }`,
  variables: {
    name: 'Clara',
  },
};

export const subscriptionGql: Subscription = {
  query: `subscription subscribeToUser($user: String){
    user(user: $user) {
      name
    }
  }
  `,
  variables: {
    user: 'colin',
  },
};

export const queryOperation: Operation = {
  key: '2',
  operationName: 'query',
  context,
  ...queryGql,
};

export const mutationOperation: Operation = {
  key: JSON.stringify(mutationGql),
  operationName: 'mutation',
  context,
  ...mutationGql,
};

export const subscriptionOperation: Operation = {
  key: JSON.stringify(subscriptionGql),
  operationName: 'subscription',
  context,
  ...subscriptionGql,
};

export const queryResponse: ExchangeResult = {
  operation: queryOperation,
  data: {
    user: {
      name: 'Clive',
    },
  },
};

export const mutationResponse: ExchangeResult = {
  operation: mutationOperation,
  data: {},
};

export const subscriptionResult: ExecutionResult = {
  data: {}
};
