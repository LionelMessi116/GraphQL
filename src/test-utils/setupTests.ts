import { configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

// Setup enzyme adapter
configure({ adapter: new Adapter() });

// Setup globals
const g = global as any;

g.AbortController = undefined;
g.fetch = jest.fn();

process.on('unhandledRejection', error => {
  throw error;
});
