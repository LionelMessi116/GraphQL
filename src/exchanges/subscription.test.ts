import { of } from 'rxjs';
import { take } from 'rxjs/operators';
import { subscriptionExchange } from './subscription';
import { subscriptionOperation } from '../test-utils';

const response = {
  data: {
    data: {
      user: 1200,
    },
  },
};

const args = {
  forward: jest.fn(),
  subject: {},
} as any;

const createOperation = res => {
  const unsubscribe = jest.fn();

  const operation = { ...subscriptionOperation };
  operation.query.replace('subscribeToUser', `${Math.random() * Date.now()}`);
  operation.context.forwardSubscription = jest.fn();

  if (res) {
    operation.context.forwardSubscription.mockImplementation((_, observer) => {
      observer.next(res);
      return { unsubscribe };
    });
  }

  return { operation, unsubscribe };
};

it('should return response data from fetch', async () => {
  const { operation } = createOperation(response);

  const data = await subscriptionExchange(args)(of(operation))
    .pipe(take(1))
    .toPromise();

  expect(data).toMatchSnapshot();
});

it('should unsubscribe from future data', async () => {
  const { operation, unsubscribe } = createOperation(response);

  // Subscribe
  await subscriptionExchange(args)(of(operation))
    .pipe(take(1))
    .toPromise();

  // Unsubscribe
  operation.context.unsubscribe = true;
  subscriptionExchange(args)(of(operation)).subscribe(() => void 0);

  expect(unsubscribe).toBeCalled();
});
