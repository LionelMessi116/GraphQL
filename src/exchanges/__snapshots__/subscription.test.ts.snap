// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should return response data from forwardSubscription observable 1`] = `
Object {
  "data": Object {},
  "error": undefined,
  "operation": Object {
    "context": Object {
      "fetchOptions": Object {
        "method": "POST",
      },
      "url": "http://localhost:3000/graphql",
    },
    "key": "{\\"query\\":\\"subscription subscribeToUser($user: String){\\\\n    user(user: $user) {\\\\n      name\\\\n    }\\\\n  }\\\\n  \\",\\"variables\\":{\\"user\\":\\"colin\\"}}",
    "operationName": "subscription",
    "query": "subscription subscribeToUser($user: String){
    user(user: $user) {
      name
    }
  }
  ",
    "variables": Object {
      "user": "colin",
    },
  },
}
`;
